### PR TITLE
ROX-25547: Create Microsoft Sentinel alert notifier configuration

### PR DESCRIPTION
**Pull request by SimonBaeumer**
### Description

This update introduces the configuration for integrating Microsoft Sentinel with various alert notifiers.

**Why Separate Configurations for StreamName and DataCollectionRuleID Are Necessary**

Each notifier source, such as _AlertNotifier_, _NetworkPolicyNotifier_, or _AuditLogNotifier_, can be associated with a different `DataCollectionRule (DCR)`. These rules define how incoming data is processed before being ingested into Azure Sentinel. Additionally, each notifier must have its own unique `StreamName`, which identifies the specific data stream associated with that notifier. Typically a `StreamName` is automatically created on table creation and matches the table's name.

To accommodate this, we create a separate configuration object for each notifier. This allows for independent configuration of both the `StreamName` and the `DataCollectionRuleID` associated with each source.

**What is a DCR?**

In Azure, logs are received as JSON objects, which must conform to a predefined schema to be properly ingested. A `Data Collection Rule (DCR)` is a crucial part of the ETL (Extract, Transform, Load) pipeline that ensures data is extracted and transformed into the correct format according to this schema.

Each log source—such as **AlertNotifier**, **NetworkPolicyNotifier**, and **AuditLogNotifier**—requires its own schema definition, which is managed by its corresponding `StreamName`. This ensures that the data from different sources is correctly processed and ingested by Azure Sentinel. Theoretically it is possible to configure the same `Data Collection Rule` for multiple schemas.

Screenshot from 2024-08-21 17-03-18

_Note: The schema doesn't reflect the current Alert schema_

**Example configuration**

Eventually this means new configurations are added, e.g. as:

```
message MicrosoftSentinel {
  // ...
  // NotifierConfig contains information about the data collection rule which is a config per notifier type.
  message NotifierConfig {
    string stream_name = 1;
    string data_collection_rule_id = 2;
  }
  NotifierConfig alert_notifier = 5;

  // added eventually...
  NotifierConfig network_policy_notifier = 6;
  NotifierConfig audit_log_notifier = 7;
}
```

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

**Comment by SimonBaeumer** [comment]
Unit tests `TestGetMetadataIntegration` is failing, I need to take a look. what's happening there. The review can still be conducted because this is about the configuration.

File: proto/storage/notifier.proto [hunk 1]
```diff
@@ -129,6 +129,31 @@ message SumoLogic {
...
+  // TODO(ROX-25739): Add client certificate
+
+  // NotifierConfig contains information about the data collection rule which is a config per notifier type.
+  message NotifierConfig {
+    string stream_name = 1;
+    string data_collection_rule_id = 2;
+  }
+  NotifierConfig alert_notifier = 5;
+
+  // TODO(ROX-25848): Add NetworkPolicyNotifier configuration
```
Comment (clickboo) [inline proto/storage/notifier.proto:153]: Do we have requirements to send network policy via this notifier? Similar for audit logs?

File: proto/storage/notifier.proto [hunk 2]
```diff
@@ -129,6 +129,31 @@ message SumoLogic {
...
+
+  // application_client_id contains the ID of the application ID of the service principal.
+  string application_client_id = 3;
+
+  // secret contains the client secret.
+  string secret = 4; // @gotags: scrub:"always"
+  // TODO(ROX-25739): Add client certificate
+
+  // NotifierConfig contains information about the data collection rule which is a config per notifier type.
+  message NotifierConfig {
```
Comment (clickboo) [inline proto/storage/notifier.proto:147]: I think you  want to use the one_of config in the main notifier proto and add a entry for microsoft sentinel.

File: proto/storage/notifier.proto (hunk 1 above)
Comment (SimonBaeumer) [reply proto/storage/notifier.proto:153]: For Audit Logs yes, it is part of the epic. For Network Policies no, it is not part of the requirements.

File: proto/storage/notifier.proto (hunk 2 above)
Comment (SimonBaeumer) [reply proto/storage/notifier.proto:147]: This is part of the next PR: https://github.com/stackrox/stackrox/pull/12480/files#diff-b4b2f26285a0a1b936166c6ba2551ba1012a44659f8c0809915e0fd4c420ad88R30

Comment (clickboo) [reply proto/storage/notifier.proto:147]: You want to rename this to AlertConfig or violation config? Given you plan to have a separate config for violations/network policy/audit log - basically the different types of data that can be sent over this notifier?

Comment (SimonBaeumer) [reply proto/storage/notifier.proto:147]: I think I was to abstract here, `DataCollectionRuelConfig` or `DCRConfig` would be more percise here. 

> Given you plan to have a separate config for violations/network policy/audit log - basically the different types of data that can be sent over this notifier?

Yes, it is required by the Log Analytics Workspace in Azure to create a schema for the logs. Each type gets a corresponding schema in Azure, that's why this must be added for each type.

Comment (SimonBaeumer) [reply proto/storage/notifier.proto]: Renamed the configuration to `DataCollectionRuleConfig`

**Review by dashrews78 — APPROVED** [review]
LGTM assuming you satisfied @clickboo

File: proto/storage/notifier.proto [hunk 3]
```diff
@@ -129,6 +129,31 @@ message SumoLogic {
...
+  string directory_tenant_id = 2;
+
+  // application_client_id contains the ID of the application ID of the service principal.
+  string application_client_id = 3;
+
+  // secret contains the client secret.
+  string secret = 4; // @gotags: scrub:"always"
+  // TODO(ROX-25739): Add client certificate
+
+  // NotifierConfig contains information about the data collection rule which is a config per notifier type.
```
Comment (clickboo) [inline proto/storage/notifier.proto:146]: nit: comment needs to change: DataCollectionRuleConfig contains ....

File: proto/storage/notifier.proto [hunk 4]
```diff
@@ -129,6 +129,31 @@ message SumoLogic {
...
+  message DataCollectionRuleConfig {
+    string stream_name = 1;
+    string data_collection_rule_id = 2;
+  }
+  DataCollectionRuleConfig alert_notifier = 5;
```
Comment (clickboo) [inline proto/storage/notifier.proto:151]: Can the same notifier have 3 different rule configs - I assume yes so this does not need to be a one_of.

If yes, food for thought: 

repeated RuleConfig rule_configs = 5;

message RuleConfig {
    oneof rule_config {
          DataCollectionRuleConfig dc_rule_config  = 1;
          NetworkPolicyRuleConfig network_config = 2;
          AuditLogRuleConfig audit_log_config = 3;
    }
}

File: proto/storage/notifier.proto (hunk 3 above)
Comment (SimonBaeumer) [reply proto/storage/notifier.proto:146]: I'll address this in a following PR

File: proto/storage/notifier.proto (hunk 4 above)
Comment (SimonBaeumer) [reply proto/storage/notifier.proto:151]: Yes, these can be three different configurations, `oneof` is not necessary

**Pull request merged**